### PR TITLE
[fix] Nullable parameter in conversion operator

### PR DESCRIPTION
**Pull request by IgorBuchelnikov**
If conversion operator contains Nullable<T> parameter:

> System.InvalidOperationException: The operands for operator 'Convert' do not match the parameters of method 'op_Explicit'.

### CODE REVIEW

**Comment by MaceWindu** [comment]
I've requested db-related test functionality removal, but want to note that:
- if we add new test objects, they should be added to db creation scripts
- it is not recommended to create new global test tables/columns without good reason

Better approach for such tests is to create local table in test itself, using CreateLocalTable test api with type-scoped mappings: [example](https://github.com/linq2db/linq2db/blob/master/Tests/Linq/UserTests/Issue1869Tests.cs#L129)

**Comment by IgorBuchelnikov** [comment]
Done. Writing test faced to problems related with inserting test row:
1) For MS Access data provider the only way to define conversion from CustomMoneyType to decimal is make CustomMoneyType implements IConvertible. That interface does not provide conversion to Nullable<decimal> (only to decimal). Moreover conversion to decimal is perfomed throuth ToString(IFormatProvider provider) member. I did not find a way to insert null value. Seems returning null from ToString(IFormatProvider provider) is treated as exclusion of parameter (not as null parameter value). 
2) For other providers conversion from CustomMoneyType to decimal behaviour is diffent. Some providers call ToDecimal(IFormatProvider provider) or GetTypeCode(). May be there is a way to insert null for somу providers.

**Comment by MaceWindu** [comment]
Rounding issue possibly cames from CreateTable generating type name without decimal part size set. This could be fixed by specifying [Scale](https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Mapping/ColumnAttribute.cs#L293) for Column attribute.

File: Tests/Linq/Common/ConvertTests.cs [hunk 1]
```diff
@@ -461,5 +463,20 @@ public void ConvertFromNullableEnum2()
 			Assert.AreEqual(0,    ConvertTo<int>. From((Enum15?)null));
 			Assert.AreEqual(null, ConvertTo<int?>.From((Enum15?)null));
 		}
+	
+		[Test]
+		public void NullableParameterInOperatorConvert([IncludeDataSources(ProviderName.Access)] string context)
+		{
+			using (var db = GetDataContext(context))
```
Comment (MaceWindu) [inline Tests/Linq/Common/ConvertTests.cs]: Why do you test against database? Convert class could be tested without db involved (which will make things easier, as you don't need to struggle with Access)

File: Tests/Model/TestDataConnection.cs [hunk 2]
```diff
@@ -65,6 +65,7 @@ void TestDataConnection_OnClosing(object sender, EventArgs e)
 		public ITable<GrandChild1>            GrandChild1            => GetTable<GrandChild1>();
 		public ITable<LinqDataTypes>          Types                  => GetTable<LinqDataTypes>();
 		public ITable<LinqDataTypes2>         Types2                 => GetTable<LinqDataTypes2>();
+		public ITable<LinqDataTypes3>         Types3                 => GetTable<LinqDataTypes3>();
```
Comment (MaceWindu) [inline Tests/Model/TestDataConnection.cs]: Don't add new members here. Better to use GetTable call in test itself.

File: Tests/Model/TestServiceModelDataContext.cs [hunk 3]
```diff
@@ -44,6 +44,7 @@ public class TestServiceModelDataContext : ServiceModelDataContext, ITestDataCon
 		public ITable<GrandChild1>            GrandChild1            { get { return this.GetTable<GrandChild1>();            } }
 		public ITable<LinqDataTypes>          Types                  { get { return this.GetTable<LinqDataTypes>();          } }
 		public ITable<LinqDataTypes2>         Types2                 { get { return this.GetTable<LinqDataTypes2>();         } }
+		public ITable<LinqDataTypes3>         Types3                 { get { return this.GetTable<LinqDataTypes3>();         } }
```
Comment (MaceWindu) [inline Tests/Model/TestServiceModelDataContext.cs]: Same here, don't extend generic test infrastructure with one-test specific functionality

File: Tests/Model/LinqDataTypes.cs [hunk 4]
```diff
@@ -136,4 +136,102 @@ public override string ToString()
 			return string.Format("{{{0,2}, {1,7}, {2:O}, {3,5}, {4}, {5}, '{6}'}}", ID, MoneyValue, DateTimeValue, BoolValue, GuidValue, SmallIntValue, StringValue);
 		}
 	}
+
+	[Table("LinqDataTypes")]
```
Comment (MaceWindu) [inline Tests/Model/LinqDataTypes.cs]: This class should belong to class with test

File: Tests/Linq/Common/ConvertTests.cs (hunk 1 above)
Comment (IgorBuchelnikov) [reply Tests/Linq/Common/ConvertTests.cs]: I tested according [examle ](https://github.com/linq2db/linq2db/blob/master/Tests/Linq/UserTests/Issue1869Tests.cs#L129) you [provided](https://github.com/linq2db/linq2db/pull/1998#issuecomment-563307005)

File: Tests/Model/TestDataConnection.cs (hunk 2 above)
Comment (IgorBuchelnikov) [reply Tests/Model/TestDataConnection.cs]: done

File: Tests/Model/TestServiceModelDataContext.cs (hunk 3 above)
Comment (IgorBuchelnikov) [reply Tests/Model/TestServiceModelDataContext.cs]: done

File: Tests/Model/LinqDataTypes.cs (hunk 4 above)
Comment (IgorBuchelnikov) [reply Tests/Model/LinqDataTypes.cs]: done

**Comment by sdanyliv** [comment]
@IgorBuchelnikov, what are you trying to achieve here? It can be easily done by configuring MappingSchema.

**Comment by IgorBuchelnikov** [comment]
[> Rounding issue possibly cames from CreateTable generating type name without decimal part size set. This could be fixed by specifying [Scale](https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Mapping/ColumnAttribute.cs#L293) for Column attribute.](https://github.com/linq2db/linq2db/pull/1998#issuecomment-564038422)

Setting of Scale attribute does't work.

File: Tests/Linq/Common/ConvertTests.cs (hunk 1 above)
Comment (MaceWindu) [reply Tests/Linq/Common/ConvertTests.cs]: `Convert<decimal?, CustomMoneyType>.From(1.11m)` should work as test code I think

**Comment by IgorBuchelnikov** [comment]
https://github.com/linq2db/linq2db/pull/1998#issuecomment-564067594

I fix linq2db internal error. ConvertBuilder could not construct expression for the test provided.

> System.InvalidOperationException: The operands for operator 'Convert' do not match the parameters of method 'op_Explicit'.

**Comment by IgorBuchelnikov** [comment]
https://github.com/linq2db/linq2db/pull/1998#issuecomment-564067594

Please, show me how to fix it by configuring MappingSchema

**Comment by sdanyliv** [comment]
Not tested, but something like this:
```cs
var ms = new MappingSchema();

ms.SetConvertExpression<decimal, CustomMoneyType>(d => new CustomMoneyType(d));

ms.SetConverter<CustomMoneyType, DataParameter>(c => 
{
     return new DataParameter(c.Amount)
})

var dc = new DataConnection(ms);
```

**Comment by sdanyliv** [comment]
Ok, i see.

**Pull request merged**